src/backend/distributed/executor/executor.c

```diff
@@ -1643,6 +1650,12 @@ ReceiveResults(WorkerSession *session, bool storeRows)
 		}
 		else if (resultStatus != PGRES_SINGLE_TUPLE)
 		{
+			/*
+			 * We can still recover from error using ROLLBACK TO SAVEPOINT,
+			 * cleanup state to allow that.
+			 */
+			FinishDistributedExecution(execution);
```

Comment:
Same comment as https://github.com/citusdata/citus/pull/2673#pullrequestreview-230549746

We should make sure you cannot roll back to a savepoint after a connection failure.

Reply:
From what I remember, @anarazel said it is not safe to error out in `ROLLBACK TO SAVEPOINT`. However postgres inserts an implicit `SAVEPOINT` command right after the rollback, which errors out if there is a connection error, so effectively not allowing rolling back to savepoints in case of connection errors.

I need to do more postgres code reading/inspection to see if we can have a better solution.

Reply:
I'm still slightly concerned about other points we could error out and continue (E.g. in `ErrorSizeLimitIsExceeded`). Any thoughts?

Reply:
ping

Reply:
Makes sense. Moved the unclaim to the `PG_CATCH()` of `RunDistributedExecution()` to cover all errors, what do you think?